### PR TITLE
Add more scenerios to rng device test

### PR DESCRIPTION
**Pull request by suqinhuang**
1. Fetch data from /dev/random in guest and host at the same time
2. Switch rng device and fetch data

Signed-off-by: Suqin Huang <shuang@redhat.com>

### CODE REVIEW

**Comment by suqinhuang** [comment]
id: 1354323 1354325

File: qemu/tests/rng_stress.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def get_rng_list(vm):
+        rng_list = []
+        for device in vm.devices:
+            if isinstance(device, qdevices.QDevice):
+                if device.get_param("driver") == "virtio-rng-pci":
+                    rng_list.append(device)
+        return rng_list
+
+    def get_current_rng(session):
+        verify_cmd = params.get("driver_verifier_cmd")
```
Comment (xutian) [inline qemu/tests/rng_stress.py]: params["driver_verifier_cmd"]  because it's mandatory cmd view from your code.

Comment (xutian) [inline qemu/tests/rng_stress.py]: doc string is required

**Comment by suqinhuang** [comment]
@pingl2016 could you help review this patch

File: qemu/tests/rng_stress.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        for rng_device in rng_devices:
+            error_context.context("Change virtio rng device to %s" %
+                                  rng_device, logging.info)
+            session.cmd_status(params.get("switch_rng_cmd") % rng_device)
+            error_context.context("Read from %s in guest" % rng_device,
+                                  logging.info)
+            utils_test.run_virt_sub_test(test, params, env, sub_test)
+
+    if params.get("post_cmd"):
+        utils.system(params.get("post_cmd"))
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: for case one_device, the read size of dd is too small to check dd process.
05:34:04 ERROR| Traceback (most recent call last):
05:34:04 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-41.0-py2.7.egg/avocado_vt/test.py", line 204, in runTest
05:34:04 ERROR|     raise exceptions.TestError(details)
05:34:04 ERROR| TestError: Command <kill -9 `ps -a -o ppid,comm | grep dd | awk '{print $1}'`> failed, rc=1, Command returned non-zero exit status    [context: Read rng device in guest]
05:34:04 ERROR| * Command:
05:34:04 ERROR|     kill -9 `ps -a -o ppid,comm | grep dd | awk '{print $1}'`
05:34:04 ERROR| Exit status: 1

File: qemu/tests/rng_stress.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            if isinstance(device, qdevices.QDevice):
+                if device.get_param("driver") == "virtio-rng-pci":
+                    rng_list.append(device)
+        return rng_list
+
+    def get_available_rng(session):
+        """
+        Get available rng devices from /sys/devices
+        """
+        verify_cmd = params.get["driver_available_cmd"]
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: should be params.get("driver_available_cmd")?

File: qemu/tests/rng_stress.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        try:
+            rng_devices = session.cmd_output_status(verify_cmd)
+        except aexpect.ShellTimeoutError:
+            err = "cat cmd timeout, pls check if it's a product bug"
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: I think it is better to show cmd here for tester to check.

File: qemu/tests/rng_stress.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    login_timeout = int(params.get("login_timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    sub_test = params.get("sub_test")
+
+    if params.get("pre_cmd"):
+        error_context.context("Fetch data from host", logging.info)
+        utils.system(params.get("pre_cmd"))
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: module utils has no import

File: qemu/tests/rng_stress.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        for rng_device in rng_devices:
+            error_context.context("Change virtio rng device to %s" %
+                                  rng_device, logging.info)
+            session.cmd_status(params.get("switch_rng_cmd") % rng_device)
+            error_context.context("Read from %s in guest" % rng_device,
+                                  logging.info)
+            utils_test.run_virt_sub_test(test, params, env, sub_test)
+
+    if params.get("post_cmd"):
+        end_time = time.time() + 20
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: module time has no import

File: qemu/tests/rng_stress.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        rng_devices = get_available_rng(session)
+        rng_attached = get_rng_list(vm)
+        if len(rng_devices) != len(rng_attached):
+            raise exceptions.TestFail("The devices get from rng_arriable"
+                                      " don't match the rng devices attached")
+
+        for rng_device in rng_devices:
+            error_context.context("Change virtio rng device to %s" %
+                                  rng_device, logging.info)
+            session.cmd_status(params.get("switch_rng_cmd") % rng_device)
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: for case one_device, switch_rng_cmd is undefined.

File: qemu/tests/rng_stress.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                    rng_list.append(device)
+        return rng_list
+
+    def get_available_rng(session):
+        """
+        Get available rng devices from /sys/devices
+        """
+        verify_cmd = params["driver_available_cmd"]
+        try:
+            rng_devices = session.cmd_output(verify_cmd)
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: After test, found the command is also contained in the output. 
06:29:40 DEBUG| Sending command: cat /sys/devices/virtual/misc/hw_random/rng_available
06:29:40 DEBUG| ###rng_devices=cat /sys/devices/virtual/misc/hw_random/rng_availab<at /sys/devices/virtual/misc/hw_random/rng_availabl                         ^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^H^He
06:29:40 DEBUG| virtio_rng.0
06:29:40 DEBUG|
06:29:40 DEBUG| ###rng_devices=['cat', '/sys/devices/virtual/misc/hw_random/rng_availab<at', '/sys/devices/virtual/misc/hw_random/rng_availabl', '\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08e', 'virtio_rng.0']
06:29:40 DEBUG| ###rng_list=[<virttest.qemu_devices.qdevices.QDevice object at 0x5a7fe10>]

File: qemu/tests/rng_stress.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                    rng_list.append(device)
+        return rng_list
+
+    def get_available_rng(session):
+        """
+        Get available rng devices from /sys/devices
+        """
+        verify_cmd = params["driver_available_cmd"]
+        try:
+            rng_devices = session.cmd_output_safe(verify_cmd)
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: cmd_output_safe() returns the command and output of the commend, so I think a regular expression is needed to get the rng devices.

File: qemu/tests/rng_stress.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        return rng_list
+
+    def get_available_rng(session):
+        """
+        Get available rng devices from /sys/devices
+        """
+        verify_cmd = params["driver_available_cmd"]
+        try:
+            output = session.cmd_output_safe(verify_cmd)
+            rng_devices = re.findall(r"virtio_rng.\d*", output)
```
Comment (pingl2016) [inline qemu/tests/rng_stress.py]: 1. module utils has no import
2. I think r"virtio_rng.\d+" is better.

**Comment by suqinhuang** [comment]
@pingl2016 hi, cfg file is updated, could you have a look

**Comment by pingl2016** [comment]
@suqinhuang It is ok for me.
(1/6) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_stress.two_device: PASS (90.42 s)
 (2/6) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.rng_stress.one_device: PASS (48.53 s)
 (3/6) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.rng_stress.two_device: PASS (59.80 s)
 (4/6) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.rng_stress.one_device: PASS (57.87 s)

**Pull request merged**